internal/print-to-source.go

```diff
@@ -0,0 +1,52 @@
+package astro
+
+import (
+	"fmt"
+	"strings"
+)
+
+func PrintToSource(buf *strings.Builder, node *Node) {
+	switch node.Type {
+	case DocumentNode:
+		for c := node.FirstChild; c != nil; c = c.NextSibling {
+			PrintToSource(buf, c)
+		}
+	case TextNode:
+		buf.WriteString(node.Data)
+	case ElementNode:
+		buf.WriteString(fmt.Sprintf(`<%s`, node.Data))
+		for _, attr := range node.Attr {
+			if attr.Namespace != "" {
+				buf.WriteString(attr.Namespace)
+				buf.WriteString(":")
+			}
+			buf.WriteString(" ")
+			switch attr.Type {
+			case QuotedAttribute:
+				buf.WriteString(attr.Key)
+				buf.WriteString("=")
+				buf.WriteString(`"` + attr.Val + `"`)
+			case EmptyAttribute:
+				buf.WriteString(attr.Key)
+			case ExpressionAttribute:
+				buf.WriteString(attr.Key)
+				buf.WriteString("=")
+				buf.WriteString(`{` + strings.TrimSpace(attr.Val) + `}`)
+			case SpreadAttribute:
+				buf.WriteString(`{...` + strings.TrimSpace(attr.Val) + `}`)
+			case ShorthandAttribute:
+				buf.WriteString(attr.Key)
+				buf.WriteString("=")
+				buf.WriteString(`{` + strings.TrimSpace(attr.Key) + `}`)
+			case TemplateLiteralAttribute:
+				buf.WriteString(attr.Key)
+				buf.WriteString("=`" + strings.TrimSpace(attr.Val) + "`")
+			}
+		}
+		buf.WriteString(`>`)
+		for c := node.FirstChild; c != nil; c = c.NextSibling {
+			PrintToSource(buf, c)
+		}
+		buf.WriteString(fmt.Sprintf(`</%s>`, node.Data))
+	}
+}
```

Comment:
Extracted this to a shared `internal/` file. Only used for tests, but I was getting an import cycle when I tried to put it in the `test_utils` package.

It's nice how straight-forward this is!